src/services/uniswap_v3.js

```diff
@@ -0,0 +1,276 @@
+    return { trade, expectedAmount }
+  }
+
+  async priceSwapOut (tokenIn, tokenOut, tokenOutAmount) {
+    await this.extend_update_pairs([tokenIn, tokenOut]);
+    const tOut = this.tokenList[tokenOut];
+    const tIn = this.tokenList[tokenIn];
+    const tokenAmountOut = new uni.TokenAmount(tOut, ethers.utils.parseUnits(tokenOutAmount, tOut.decimals));
+    if (this.pairs.length === 0){
+      const route = await this.fetch_route(tIn, tOut);
+      const trade = uni.Trade.exactOut(route, tokenAmountOut);
+      if ( trade !== undefined ){
+        const expectedAmount = trade.maximumAmountIn(this.allowedSlippage);
+        this.cachedRoutes[tIn.symbol + tOut.Symbol] = trade;
+        return { trade, expectedAmount }
+      }
+      return
+    }
+    const trade = uni.Trade.bestTradeExactOut(this.pairs, this.tokenList[tokenIn], tokenAmountOut, { maxHops: 5 })[0];
+    if (trade === undefined){trade = this.cachedRoutes[tIn.symbol + tOut.Symbol];}
+    else{this.cachedRoutes[tIn.symbol + tOut.Symbol] = trade;}
+    const expectedAmount = trade.maximumAmountIn(this.allowedSlippage);
+    return { trade, expectedAmount }
+  }
+
+  async swapExactIn (wallet, trade, tokenAddress, gasPrice) {
+    const result = uni.Router.swapCallParameters(
+      trade,
+      {
+        ttl: TTL,
+        recipient: wallet.address,
+        allowedSlippage: this.allowedSlippage
+      }
+    )
+
+    const contract = new ethers.Contract(this.router, routerArtifact.abi, wallet)
+    const tx = await contract.[result.methodName](
+      ...result.args,
+      {
+        gasPrice: gasPrice * 1e9,
+        gasLimit: GAS_LIMIT,
+        value: result.value
+      }
+    )
+
+    debug(`Tx Hash: ${tx.hash}`);
+    return tx
+  }
+
+  async swapExactOut (wallet, trade, tokenAddress, gasPrice) {
+    const result = uni.Router.swapCallParameters(
+      trade,
+      {
+        ttl: TTL,
+        recipient: wallet.address,
+        allowedSlippage: this.allowedSlippage
+      }
+    )
+
+    const contract = new ethers.Contract(this.router, routerArtifact.abi, wallet)
+    const tx = await contract.[result.methodName](
+      ...result.args,
+      {
+        gasPrice: gasPrice * 1e9,
+        gasLimit: GAS_LIMIT,
+        value: result.value
+      }
+    )
+
+    debug(`Tx Hash: ${tx.hash}`);
+    return tx
+  }
+
+  // LP section
+
+  async getPosition (wallet, tokenId) {
+    const contract = new ethers.Contract(this.router, routerArtifact.abi, wallet);
+    return await contract.positions(tokenId);
+  }
+
+  async addPosition (wallet, token0, token1, amount0, amount1, fee, tick0, tick1) {
+    const contract = new ethers.Contract(this.router, routerArtifact.abi, wallet);
+    const priceSqrt = require('./uniswap-v3/encodePriceSqrt');
+    const initPool = await contract.createAndInitializePoolIfNecessary(
+        token0.address,
+        token1.address,
+        FeeAmount.MEDIUM,
+        encodePriceSqrt(1, 1)
+      );
+    console.log(initPool);
```

Comment:
pls not leave console.log what you used for debugging in code

Reply:
Thanks for catching again!